compiler/src/iree/compiler/Codegen/SPIRV/SPIRVEmulateI64.cpp

```diff
@@ -0,0 +1,200 @@
+    Type newResTy = getTypeConverter()->convertType(op.getType());
+    if (!newResTy)
+      return rewriter.notifyMatchFailure(op->getLoc(),
+                                         "failed to legalize memref type");
+
+    auto newOp = rewriter.replaceOpWithNewOp<memref::LoadOp>(
+        op, newResTy, adaptor.getMemref(), adaptor.getIndices());
+    LLVM_DEBUG(llvm::dbgs()
+               << "WideIntegerEmulation: new op: " << newOp << "\n");
+    (void)newOp;
+    return success();
+  }
+};
+
+// TODO(kuhar): Upstream memref conversion patterns.
+struct ConvertMemrefStore final : OpConversionPattern<memref::StoreOp> {
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult matchAndRewrite(
+      memref::StoreOp op, OpAdaptor adaptor,
+      ConversionPatternRewriter &rewriter) const override {
+    Type newTy = getTypeConverter()->convertType(op.getMemRefType());
+    if (!newTy)
+      return rewriter.notifyMatchFailure(op->getLoc(),
+                                         "failed to legalize memref type");
+
+    auto newOp = rewriter.replaceOpWithNewOp<memref::StoreOp>(
+        op, adaptor.getValue(), adaptor.getMemref(), adaptor.getIndices());
+    LLVM_DEBUG(llvm::dbgs()
+               << "WideIntegerEmulation: new op: " << newOp << "\n");
+    (void)newOp;
+    return success();
+  }
+};
+
+//===----------------------------------------------------------------------===//
+// Helper functions
+//===----------------------------------------------------------------------===//
+
+// TODO(kuhar): Upstream memref conversion patterns.
+static void populateMemrefEmulationPatterns(
+    arith::WideIntEmulationConverter &converter, RewritePatternSet &patterns) {
+  patterns.add<ConvertMemrefLoad, ConvertMemrefStore>(converter,
+                                                      patterns.getContext());
+}
+
+// TODO(kuhar): Upstream memref conversion patterns.
+static void populateMemrefWideIntegerEmulationConversions(
+    arith::WideIntEmulationConverter &typeConverter) {
+  typeConverter.addConversion(
+      [&typeConverter](MemRefType ty) -> Optional<Type> {
+        auto intTy = ty.getElementType().dyn_cast<IntegerType>();
+        if (!intTy) return ty;
+
+        if (intTy.getIntOrFloatBitWidth() <=
+            typeConverter.getMaxTargetIntBitWidth())
+          return ty;
+
+        Type newElemTy = typeConverter.convertType(intTy);
+        if (!newElemTy) return None;
+
+        return ty.cloneWith(None, newElemTy);
+      });
+}
+
+static void populateIreeI64EmulationPatterns(
+    arith::WideIntEmulationConverter &converter, RewritePatternSet &patterns) {
+  patterns.add<ConvertHalInterfaceBindingSubspan>(converter,
+                                                  patterns.getContext());
+}
+
+static bool supportsI64(ModuleOp op) {
+  spirv::TargetEnvAttr targetEnv = getSPIRVTargetEnvAttr(op);
+  assert(targetEnv && "Not a valid spirv module");
+  return llvm::is_contained(targetEnv.getCapabilities(),
```

Comment:
We should create a `spirv::TargetEnv` out of the `spirv::TargetEnvAttr` to use the `allows` API there. The difference is, `spirv::TargetEnvAttr` only gives us the exact same set of capabilities specified by the environment, but certain capabilities are _implied_ (e.g., `Matrix` is always implied with `Shader`). `spirv::TargetEnv` handles the implication relationship under the hood.